src/NodaTime/DateInterval.cs

```diff
@@ -214,6 +214,31 @@ public DateInterval Intersection([NotNull]DateInterval interval)
                 : null;
         }
 
+        /// <summary>
+        /// Returns the union between the given interval and this interval, as long as they're overlapping or contiguous.
+        /// </summary>
+        /// <param name="interval">
+        /// The specified interval to unite with this one. 
+        /// </param>
+        /// <returns>
+        /// A <see cref="DateInterval"/> corresponding to the union between the given interval and the current
+        /// instance, in the case the intervals overlap or are contiguous; a null reference otherwise.
+        /// </returns>
+        /// <exception cref="ArgumentException"><paramref name="interval" /> uses a different
+        /// calendar to this date interval.</exception>
+        [CanBeNull]
+        public DateInterval Union([NotNull] DateInterval interval)
+        {
+            ValidateInterval(interval);
+
+            var start = LocalDate.Min(Start, interval.Start);
+            var end = LocalDate.Max(End, interval.End);
+
+            return Period.Between(start, end).Days >= Length + interval.Length
```

Comment:
It's try to avoid using Period.Between here - that could do quite a bit of arithmetic.

I'll have a think about the best way of detecting this.

Reply:
And actually, this will be incorrect, as it will fail if there are multiple years and months in the period - you'd need to specify PeriodUnits.Days. At that point, that's actually reasonably cheap (IIRC) - so maybe that's the right approach to take.

(We may want to write an internal Period.DaysBetween method to make this and Length even cheaper, but that's a different matter.)

Reply:
Having looked at this, feel free to use Period.Between but making sure you specify PeriodUnits.Days, and I'll fix it up to be more efficient later.

Before you change it to be right though, it would be good to write a test that shows it's broken: I suspect if you create two date intervals for the same dates but different years, then the union will be non-null at the moment, but should be null.